activerecord/lib/active_record/connection_adapters/postgresql/result.rb

```diff
@@ -0,0 +1,112 @@
+# frozen_string_literal: true
+
+module ActiveRecord::ConnectionAdapters::PostgreSQL
+  ###
+  # This class encapsulates a result returned from calling
+  # {#exec_query}[rdoc-ref:ConnectionAdapters::DatabaseStatements#exec_query]
+  # on the PostgreSQL connection adapter.
+  class Result < ActiveRecord::Result
+    include Enumerable
+
+    attr_reader :columns
+
+    def initialize(pg_result, adapter)
+      @pg_result = pg_result
+      @adapter = adapter
+      @field_map = nil
+      @length = nil
+      @columns = @pg_result.fields
+    end
+
+    def initialize_copy(other)
+    end
+
+    def each
+      return to_enum(:each) { length } unless block_given?
+
+      l = length
+      i = 0
+      while i < l
+        yield @pg_result.tuple(i)
+        i += 1
+      end
+    end
```

Comment:
`PostgreSQL::Result` doesn't return `Hash` but `PG::Tuple` on `each`, `[]`, `first` and `last` methods. `PG::Tuple` implements enough Hash-like functions to be used as a `Hash` replacement in ActiveRecord. It allows also Array like access. See: https://deveiate.org/code/pg/PG/Tuple.html